app/services/course/user_registration_service.rb

```diff
@@ -5,16 +5,33 @@ class Course::UserRegistrationService
   # @return [bool] True if the registration succeeded. False if the registration failed.
   def register(registration)
     CourseUser.transaction do
-      if registration.code.empty?
-        register_course_user(registration)
+      course_user, success = create_or_update_registration(registration)
+      if success && course_user.requested?
+        notify_course_staff(registration.course, course_user)
       else
-        claim_registration_code(registration)
+        success
       end
     end
   end
 
   private
 
+  # Creates the effect of performing the given registration.
+  #
+  # @param [Course::Registration] registration The registration object to be processed.
+  # @return [Array<(CourseUser, bool)>] A tuple containing the Course User which was created or
+  #   updated from the registration, as well as the success in saving the Course User.
+  def create_or_update_registration(registration)
```

Comment:
To be consistent, is it better to just return the `user_course` when success, and return `nil/false` when failed ? 

Return two different kinds of things here is confusing.